diff -ruN a/png.c b/png.c
--- a/png.c	2025-12-06 00:46:38.000000000 +0300
+++ b/png.c	2026-01-04 05:43:34.724345000 +0300
@@ -139,7 +139,7 @@
 png_reset_crc(png_structrp png_ptr)
 {
    /* The cast is safe because the crc is a 32-bit value. */
-   png_ptr->crc = (png_uint_32)crc32(0, Z_NULL, 0);
+   png_ptr->crc = (png_uint_32)zng_crc32(0, Z_NULL, 0);
 }
 
 /* Calculate the CRC over a section of data.  We can only pass as
@@ -182,7 +182,7 @@
             safe_length = (uInt)-1; /* evil, but safe */
 #endif
 
-         crc = crc32(crc, ptr, safe_length);
+         crc = zng_crc32(crc, ptr, safe_length);
 
          /* The following should never issue compiler warnings; if they do the
           * target system has characteristics that will probably violate other
@@ -979,7 +979,7 @@
       return Z_STREAM_ERROR;
 
    /* WARNING: this resets the window bits to the maximum! */
-   return inflateReset(&png_ptr->zstream);
+   return zng_inflateReset(&png_ptr->zstream);
 }
 #endif /* READ */
 
diff -ruN a/pnglibconf.h b/pnglibconf.h
--- a/pnglibconf.h	2025-12-14 15:07:27.142035400 +0300
+++ b/pnglibconf.h	2026-01-04 05:46:48.387592600 +0300
@@ -223,7 +223,6 @@
 #define PNG_USER_HEIGHT_MAX 1000000
 #define PNG_USER_WIDTH_MAX 1000000
 #define PNG_ZBUF_SIZE 8192
-#define PNG_ZLIB_VERNUM 0x1310
 #define PNG_Z_DEFAULT_COMPRESSION 9
 #define PNG_Z_DEFAULT_NOFILTER_STRATEGY 0
 #define PNG_Z_DEFAULT_STRATEGY 1
diff -ruN a/pngpriv.h b/pngpriv.h
--- a/pngpriv.h	2025-12-06 00:46:38.000000000 +0300
+++ b/pngpriv.h	2026-01-04 05:47:20.483113800 +0300
@@ -1020,22 +1020,6 @@
 #include "pngstruct.h"
 #include "pnginfo.h"
 
-/* Validate the include paths - the include path used to generate pnglibconf.h
- * must match that used in the build, or we must be using pnglibconf.h.prebuilt:
- */
-#if PNG_ZLIB_VERNUM != 0 && PNG_ZLIB_VERNUM != ZLIB_VERNUM
-#  error The include path of <zlib.h> is incorrect
-   /* When pnglibconf.h was built, the copy of zlib.h that it used was not the
-    * same as the one being used here.  Considering how libpng makes decisions
-    * to use the zlib API based on the zlib version number, the -I options must
-    * match.
-    *
-    * A possible cause of this mismatch is that you passed an -I option in
-    * CFLAGS, which is unlikely to work.  All the preprocessor options, and all
-    * the -I options in particular, should be in CPPFLAGS.
-    */
-#endif
-
 /* This is used for 16-bit gamma tables -- only the top level pointers are
  * const; this could be changed:
  */
@@ -1589,13 +1573,9 @@
 /* Initialize the row buffers, etc. */
 PNG_INTERNAL_FUNCTION(void,png_read_start_row,(png_structrp png_ptr),PNG_EMPTY);
 
-#if ZLIB_VERNUM >= 0x1240
 PNG_INTERNAL_FUNCTION(int,png_zlib_inflate,(png_structrp png_ptr, int flush),
       PNG_EMPTY);
 #  define PNG_INFLATE(pp, flush) png_zlib_inflate(pp, flush)
-#else /* Zlib < 1.2.4 */
-#  define PNG_INFLATE(pp, flush) inflate(&(pp)->zstream, flush)
-#endif /* Zlib < 1.2.4 */
 
 #ifdef PNG_READ_TRANSFORMS_SUPPORTED
 /* Optional call to update the users info structure */
diff -ruN a/pngread.c b/pngread.c
--- a/pngread.c	2025-12-06 00:46:38.000000000 +0300
+++ b/pngread.c	2026-01-04 05:44:02.196584300 +0300
@@ -795,7 +795,7 @@
    png_ptr->free_me &= ~PNG_FREE_TRNS;
 #endif
 
-   inflateEnd(&png_ptr->zstream);
+   zng_inflateEnd(&png_ptr->zstream);
 
 #ifdef PNG_PROGRESSIVE_READ_SUPPORTED
    png_free(png_ptr, png_ptr->save_buffer);
diff -ruN a/pngrutil.c b/pngrutil.c
--- a/pngrutil.c	2025-07-01 23:50:00.000000000 +0300
+++ b/pngrutil.c	2026-01-04 05:49:21.506248200 +0300
@@ -446,7 +446,6 @@
     */
    {
       int ret; /* zlib return code */
-#if ZLIB_VERNUM >= 0x1240
       int window_bits = 0;
 
 # if defined(PNG_SET_OPTION_SUPPORTED) && defined(PNG_MAXIMUM_INFLATE_WINDOW)
@@ -463,7 +462,6 @@
       }
 # endif
 
-#endif /* ZLIB_VERNUM >= 0x1240 */
 
       /* Set this for safety, just in case the previous owner left pointers to
        * memory allocations.
@@ -475,20 +473,12 @@
 
       if ((png_ptr->flags & PNG_FLAG_ZSTREAM_INITIALIZED) != 0)
       {
-#if ZLIB_VERNUM >= 0x1240
-         ret = inflateReset2(&png_ptr->zstream, window_bits);
-#else
-         ret = inflateReset(&png_ptr->zstream);
-#endif
+         ret = zng_inflateReset2(&png_ptr->zstream, window_bits);
       }
 
       else
       {
-#if ZLIB_VERNUM >= 0x1240
-         ret = inflateInit2(&png_ptr->zstream, window_bits);
-#else
-         ret = inflateInit(&png_ptr->zstream);
-#endif
+         ret = zng_inflateInit2(&png_ptr->zstream, window_bits);
 
          if (ret == Z_OK)
             png_ptr->flags |= PNG_FLAG_ZSTREAM_INITIALIZED;
@@ -514,7 +504,6 @@
 #endif
 }
 
-#if ZLIB_VERNUM >= 0x1240
 /* Handle the start of the inflate stream if we called inflateInit2(strm,0);
  * in this case some zlib versions skip validation of the CINFO field and, in
  * certain circumstances, libpng may end up displaying an invalid image, in
@@ -535,9 +524,8 @@
       png_ptr->zstream_start = 0;
    }
 
-   return inflate(&png_ptr->zstream, flush);
+   return zng_inflate(&png_ptr->zstream, flush);
 }
-#endif /* Zlib >= 1.2.4 */
 
 #ifdef PNG_READ_COMPRESSED_TEXT_SUPPORTED
 #if defined(PNG_READ_zTXt_SUPPORTED) || defined (PNG_READ_iTXt_SUPPORTED)
@@ -721,7 +709,7 @@
              * with Z_FINISH in almost all cases, so the window will not be
              * maintained.
              */
-            if (inflateReset(&png_ptr->zstream) == Z_OK)
+            if (zng_inflateReset(&png_ptr->zstream) == Z_OK)
             {
                /* Because of the limit checks above we know that the new,
                 * expanded, size will fit in a size_t (let alone an
diff -ruN a/pngstruct.h b/pngstruct.h
--- a/pngstruct.h	2025-12-06 00:46:38.000000000 +0300
+++ b/pngstruct.h	2026-01-04 05:52:11.138291600 +0300
@@ -16,30 +16,10 @@
 
 #ifndef PNGSTRUCT_H
 #define PNGSTRUCT_H
-/* zlib.h defines the structure z_stream, an instance of which is included
- * in this structure and is required for decompressing the LZ compressed
- * data in PNG files.
- */
-#ifndef ZLIB_CONST
-   /* We must ensure that zlib uses 'const' in declarations. */
-#  define ZLIB_CONST
-#endif
-#include "zlib.h"
-#ifdef const
-   /* zlib.h sometimes #defines const to nothing, undo this. */
-#  undef const
-#endif
-
-/* zlib.h has mediocre z_const use before 1.2.6, this stuff is for compatibility
- * with older builds.
- */
-#if ZLIB_VERNUM < 0x1260
-#  define PNGZ_MSG_CAST(s) png_constcast(char*,s)
-#  define PNGZ_INPUT_CAST(b) png_constcast(png_bytep,b)
-#else
+#include "zlib-ng.h"
+
 #  define PNGZ_MSG_CAST(s) (s)
 #  define PNGZ_INPUT_CAST(b) (b)
-#endif
 
 /* zlib.h declares a magic type 'uInt' that limits the amount of data that zlib
  * can handle at once.  This type need be no larger than 16 bits (so maximum of
@@ -159,7 +139,7 @@
    png_uint_32 transformations; /* which transformations to perform */
 
    png_uint_32 zowner;        /* ID (chunk type) of zstream owner, 0 if none */
-   z_stream    zstream;       /* decompression structure */
+   zng_stream    zstream;       /* decompression structure */
 
 #ifdef PNG_WRITE_SUPPORTED
    png_compression_bufferp zbuffer_list; /* Created on demand during write */
@@ -247,9 +227,7 @@
                               /* pixel depth used for the row buffers */
    png_byte transformed_pixel_depth;
                               /* pixel depth after read/write transforms */
-#if ZLIB_VERNUM >= 0x1240
    png_byte zstream_start;    /* at start of an input zlib stream */
-#endif /* Zlib >= 1.2.4 */
 #if defined(PNG_READ_FILLER_SUPPORTED) || defined(PNG_WRITE_FILLER_SUPPORTED)
    png_uint_16 filler;           /* filler bytes for pixel expansion */
 #endif
diff -ruN a/pngwrite.c b/pngwrite.c
--- a/pngwrite.c	2025-12-06 00:46:38.000000000 +0300
+++ b/pngwrite.c	2026-01-04 05:49:39.810614000 +0300
@@ -988,7 +988,7 @@
 
    /* Free any memory zlib uses */
    if ((png_ptr->flags & PNG_FLAG_ZSTREAM_INITIALIZED) != 0)
-      deflateEnd(&png_ptr->zstream);
+       zng_deflateEnd(&png_ptr->zstream);
 
    /* Free our memory.  png_free checks NULL for us. */
    png_free_buffer_list(png_ptr, &png_ptr->zbuffer_list);
diff -ruN a/pngwutil.c b/pngwutil.c
--- a/pngwutil.c	2025-04-30 16:51:46.000000000 +0300
+++ b/pngwutil.c	2026-01-04 05:50:46.946965400 +0300
@@ -409,7 +409,7 @@
          png_ptr->zlib_set_mem_level != memLevel ||
          png_ptr->zlib_set_strategy != strategy))
       {
-         if (deflateEnd(&png_ptr->zstream) != Z_OK)
+         if (zng_deflateEnd(&png_ptr->zstream) != Z_OK)
             png_warning(png_ptr, "deflateEnd failed (ignored)");
 
          png_ptr->flags &= ~PNG_FLAG_ZSTREAM_INITIALIZED;
@@ -427,11 +427,11 @@
        * do a simple reset to the previous parameters.
        */
       if ((png_ptr->flags & PNG_FLAG_ZSTREAM_INITIALIZED) != 0)
-         ret = deflateReset(&png_ptr->zstream);
+         ret = zng_deflateReset(&png_ptr->zstream);
 
       else
       {
-         ret = deflateInit2(&png_ptr->zstream, level, method, windowBits,
+         ret = zng_deflateInit2(&png_ptr->zstream, level, method, windowBits,
              memLevel, strategy);
 
          if (ret == Z_OK)
@@ -594,7 +594,7 @@
          }
 
          /* Compress the data */
-         ret = deflate(&png_ptr->zstream,
+         ret = zng_deflate(&png_ptr->zstream,
              input_len > 0 ? Z_NO_FLUSH : Z_FINISH);
 
          /* Claw back input data that was not consumed (because avail_in is
@@ -995,7 +995,7 @@
       png_ptr->zstream.avail_in = avail;
       input_len -= avail;
 
-      ret = deflate(&png_ptr->zstream, input_len > 0 ? Z_NO_FLUSH : flush);
+      ret = zng_deflate(&png_ptr->zstream, input_len > 0 ? Z_NO_FLUSH : flush);
 
       /* Include as-yet unconsumed input */
       input_len += png_ptr->zstream.avail_in;
